File: fs/src/main/scala/farjs/fs/FSServices.scala
package farjs.fs

import farjs.fs.popups.FolderShortcutsService
import scommons.react.raw.NativeContext

import scala.scalajs.js
import scala.scalajs.js.annotation.JSImport

trait FSServices extends js.Object {

  val folderShortcuts: FolderShortcutsService
}

@js.native
@JSImport("../fs/FSServices.mjs", JSImport.Default)
object FSServices extends js.Object {

  val Context: NativeContext = js.native
  
  def useServices(): FSServices = js.native
}

File: fs/FSServices.mjs
/**
 * @import { FolderShortcutsService } from "./popups/FolderShortcutsService.mjs"
 */
import React, { useContext } from "react";

const FSServices = {
  Context: React.createContext(/** @type {FSServices | null} */ (null)),

  /**
   * @returns {FSServices}
   */
  useServices: () => {
    const ctx = useContext(FSServices.Context);
    if (!ctx) {
      throw Error(
        "FSServices.Context is not found." +
          "\nPlease, make sure you use FSServices.Context.Provider in parent components",
      );
    }
    return ctx;
  },
};

export default FSServices;

/**
 * @typedef {{
 *  readonly folderShortcuts: FolderShortcutsService;
 * }} FSServices
 */
